### DISCUSSION TITLE
random BN specifications

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#24] [Q&A]
Hi, is there a way to randomly generate BN with the ability to make half of the nodes binary and half of the nodes have 20 states for example? or picking uniformly at random to assign different numbers of states on top of `randomBN` ?

**Comment by phwuil** [comment]
Hi @kenneth-lee-ch ,

both `gum.BNGenerator` and the shortcuts `gum.randomBN` can assign different numbers of states. The domain_size you give is just the max.

![image](https://github.com/agrumery/aGrUM/assets/5980636/99f892f3-99bc-4150-ab72-8422c074bb0e)

 and
 
 
![image](https://github.com/agrumery/aGrUM/assets/5980636/4b09a6d1-86d5-45ec-91ae-5fe5b433b28c)

**Reply by kenneth-lee-ch** [reply]
But this won't let me to specifically have half of the nodes being 2 states and half of them being 20 states, right?

**Reply by phwuil** [reply]
Yes you are right, this is a specific demand and I guess you have to do it with a script. For instance :

```python
import random

def randomBNForKennethLee(nodes:int,arcs:int,mod1:int=2,mod2:int=20):
  g=gum.BNGenerator()
  tmp=g.generate(nodes,arcs,2)
  
  bn=gum.BayesNet()
  # Nodes
  v=list(tmp.names())
  random.shuffle(v)
  h=len(v)//2
  for name in v[:h]:
    bn.add(name,mod1)
  for name in v[h:]:
    bn.add(name,mod2)
    
  # arcs
  bn.beginTopologyTransformation()
  for a,b in tmp.arcs():
    bn.addArc(tmp.variable(a).name(),tmp.variable(b).name())
  bn.endTopologyTransformation()
  bn.generateCPTs()
  
  return bn 
```

![image](https://github.com/agrumery/aGrUM/assets/5980636/eebcedd0-3385-4e0d-8375-148240971ede)

**Reply by kenneth-lee-ch** [reply]
@phwuil  Thank you so much! I love the function name. You are the best.

**Reply by kenneth-lee-ch** [reply]
@phwuil I tried to generate DAGs that have 80% of the variables as binary variables and 20% of them to have 80 states, I set the number of variables to be 15 with 70 edges. I often ran into the following memory errors, do you know why? It was working when I :
'''
return _pyAgrum.BayesNet_endTopologyTransformation(self)
Exception: std::bad_alloc
'''

Here is the code I used:
```

def randomBNwithSpecificStates(nodes:int,arcs:int, states: List[int], p:float, graphidx):
    table = {}
    gum.initRandom(graphidx)
    g=gum.BNGenerator()
    tmp=g.generate(nodes,arcs,2)
    bn=gum.BayesNet()
    # Nodes
    v=list(tmp.names())
    random.shuffle(v)
    # h=len(v)//2
    for name in v:
        s = np.random.choice(a=np.array(states), size=1, p=p)
        state_num = s[0]
        bn.add(name, int(state_num))
        id = bn.ids(set([name]))
        table[id[0]] = state_num
    
    # arcs
    bn.beginTopologyTransformation()
    for a,b in tmp.arcs():
        bn.addArc(tmp.variable(a).name(),tmp.variable(b).name())
    bn.endTopologyTransformation()
    bn.generateCPTs()
    # output_dict = {value: key for key, value in table.items()}
    print(table)
    return bn, table

n = 30
num_edges = 70
for i in range(100):
       randomBNwithSpecificStates(n, num_edges, states =[2, 80], p=[0.2, 0.8], graphidx=i)
```

**Reply by phwuil** [reply]
The moment where you `endTopologyTransformation` is the moment where all the CPTs are created. It may happen that when you created your graph, unfortunately, a 80-states variable is the children of many 80-state other variables. This leads to a massive amount of memory needed for this CPT. 

- 3 80-states parents for an 80-state variable means 8*80^4 = 312 Mo for this variable only.
- 4 80-states parents for an 80-state variable is more than 20Go for this variable only.

There is nothing to do about it : `randomBNwithSpecificStates` may lead to BN bigger than your memory. You have to change this algorithm... by restricting the number of parents (especially) with 80-states, for instance.

**Reply by kenneth-lee-ch** [reply]
@phwuil  do you know how to restrict the number of parents in this case?

**Reply by phwuil** [reply]
yes : you have to modify tmp before creating bn in your code : fon instance, each time a node has too many parents, randomly remove a parent.
